rclcpp/include/rclcpp/qos.hpp

```diff
@@ -140,6 +148,31 @@ class RCLCPP_PUBLIC QoS
   const rmw_qos_profile_t &
   get_rmw_qos_profile() const;
 
+  /// Check if this QoS profile is compatible with another.
+  /**
+   * Two QoS profiles are compatible if a publisher and subcription
+   * using the QoS policies can communicate with each other.
+   *
+   * This QoS profile is assumed to be for a subscription.
+   *
+   * If there is an unknown policy value in either profile, then an exception is thrown.
+   *
+   * \param[in] publisher_qos: The QoS profile for a publisher.
+   * \param[out] reason: An explanation for any incompatibility.
+   *   Set if the return value is QoSProfileCompatibility::Warning or
+   *   QoSProfileCompatibility::Error.
+   *   Not set if the QoS profiles are compatible.
+   *   This parameter is optional.
+   * \return QoSProfileCompatibility::Ok if the QoS profiles are compatible, or
+   * \return QoSProfileCompatibility::Warning if there is a chance the QoS profiles are not
+   *   compatible (this can happen if some policies are set ot "system default"), or
+   * \return QoSProfileCompatibility::Error if the QoS profiles are not compatible.
+   * \throws rclcpp::exceptions::QoSCheckCompatibilityException if any policy value is "unknown",
+   *   or if an unexpected error occurs.
+   */
+  QoSProfileCompatibility
+  compatible(const QoS & publisher_qos, std::string * reason = nullptr);
```

Comment:
For c++, it might be more natural to return a structure that contains both the Ok/Warning/Error and the reason. But this interface isn't wrong or anything, just a bit cumbersome.

Reply:
Now returning a struct 92a1cef